source/extensions/filters/common/attributes/attributes.cc

```diff
@@ -0,0 +1,551 @@
+  case UpstreamToken::SUBJECT_PEER_CERTIFICATE:
+    return ValueUtil::stringValue(upstreamSsl->subjectPeerCertificate());
+  case UpstreamToken::DNS_SAN_LOCAL_CERTIFICATE:
+    return ValueUtil::stringValue(upstreamSsl->dnsSansLocalCertificate().front());
+  case UpstreamToken::DNS_SAN_PEER_CERTIFICATE:
+    return ValueUtil::stringValue(upstreamSsl->dnsSansPeerCertificate().front());
+  case UpstreamToken::URI_SAN_LOCAL_CERTIFICATE:
+    return ValueUtil::stringValue(upstreamSsl->uriSanLocalCertificate().front());
+  case UpstreamToken::URI_SAN_PEER_CERTIFICATE:
+    return ValueUtil::stringValue(upstreamSsl->uriSanPeerCertificate().front());
+  }
+  return ValueUtil::nullValue();
+}
+
+Value* Attributes::get(ConnectionToken tok) {
+  // Get the tokens out of the way that do not need access to the downstreamSslConnection.
+  if (tok == ConnectionToken::ID) {
+    auto id = stream_info_.downstreamAddressProvider().connectionID();
+    if (id.has_value()) {
+      return ValueUtil::uint64Value(id.value());
+    } else {
+      return ValueUtil::nullValue();
+    }
+  } else if (tok == ConnectionToken::REQUESTED_SERVER_NAME) {
+    auto sn = stream_info_.downstreamAddressProvider().requestedServerName();
+    return ValueUtil::stringValue(std::string(sn));
+  }
+
+  auto downstreamSsl = stream_info_.downstreamSslConnection();
+  if (downstreamSsl == nullptr) {
+    return ValueUtil::nullValue();
+  }
+
+  std::cerr << "before switch\n";
+  switch (tok) {
+  // handled above, unreachable
+  case ConnectionToken::ID:
+  case ConnectionToken::REQUESTED_SERVER_NAME:
+    break;
+
+  case ConnectionToken::MTLS:
+    return ValueUtil::boolValue(downstreamSsl->peerCertificatePresented());
+  case ConnectionToken::TLS_VERSION:
+    return ValueUtil::stringValue(downstreamSsl->tlsVersion());
+  case ConnectionToken::SUBJECT_LOCAL_CERTIFICATE:
+    return ValueUtil::stringValue(downstreamSsl->subjectLocalCertificate());
+  case ConnectionToken::SUBJECT_PEER_CERTIFICATE:
+    return ValueUtil::stringValue(downstreamSsl->subjectPeerCertificate());
+  case ConnectionToken::DNS_SAN_LOCAL_CERTIFICATE:
+    return ValueUtil::stringValue(std::string(downstreamSsl->dnsSansLocalCertificate().front()));
+  case ConnectionToken::DNS_SAN_PEER_CERTIFICATE:
+    return ValueUtil::stringValue(std::string(downstreamSsl->dnsSansPeerCertificate().front()));
+  case ConnectionToken::URI_SAN_LOCAL_CERTIFICATE:
+    return ValueUtil::stringValue(std::string(downstreamSsl->uriSanLocalCertificate().front()));
+  case ConnectionToken::URI_SAN_PEER_CERTIFICATE:
+    return ValueUtil::stringValue(std::string(downstreamSsl->uriSanPeerCertificate().front()));
+  case ConnectionToken::TERMINATION_DETAILS:
+    return ValueUtil::optionalStringValue(stream_info_.connectionTerminationDetails());
+  }
+
+  return ValueUtil::nullValue();
+}
+
+Value* Attributes::getMetadata() {
+  //   if (attributes_.contains(METADATA_TOKEN)) {
```

Comment:
Is this something that we don't need, or that we haven't figured out yet?

Reply:
This was crashing for me in manual testing but I've since sorted out that crash, I just need to uncomment and write tests around this.